src/DurableTask.Core/Common/Utils.cs

```diff
@@ -512,5 +516,111 @@ public static bool TryGetTaskScheduledId(HistoryEvent historyEvent, out int task
                     return false;
             }
         }
+
+        /// <summary>
+        /// Gets the generic return type for a specific <paramref name="methodInfo"/>.
+        /// </summary>
+        /// <param name="methodInfo">The method to get the generic return type for.</param>
+        /// <param name="genericArguments">The generic method arguments.</param>
+        internal static Type GetGenericReturnType(MethodInfo methodInfo, Type[] genericArguments)
+        {
+            if (!methodInfo.ReturnType.IsGenericType)
+            {
+                throw new Exception("Return type is not a generic type. Type Name: " + methodInfo.ReturnType.FullName);
+            }
+
+            var genericArgument = methodInfo.ReturnType.GetGenericArguments().SingleOrDefault() ??
+                throw new Exception("Generic Parameters are not equal to 1. Type Name: " + methodInfo.ReturnType.FullName);
```

Comment:
I'm not sure what this exception message is trying to communicate to the user. Can we put it in more actionable terms? Also, can we use a more specific exception type? If I understand correctly, I wonder if the following would be clearer:
```suggestion
                throw new NotSupportedException($"The method {methodInfo.Name} cannot be used because its return type '{methodInfo.ReturnType.FullName}' has more than one generic parameter.");
```

Reply:
Agreed. This part was moved from existing code. Updating.